### PR TITLE
Implement Kafka Write Schema Transform

### PR DESCRIPTION
**Pull request by pabloem**
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by pabloem** [comment]
r: @Abacn

**Comment by pabloem** [comment]
Run Java PreCommit

**Comment by Abacn** [comment]
Thanks! Looking into it

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaReadSchemaTransformProvider.java [hunk 1]
```diff
@@ -68,15 +81,22 @@ public List<String> inputCollectionNames() {
 
   @Override
   public List<String> outputCollectionNames() {
-    return Lists.newArrayList("OUTPUT");
+    return Lists.newArrayList("output");
```
Comment (Abacn) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaReadSchemaTransformProvider.java:84]: Just wondering the purpose of UPPER to lower case change. Is it some cleanup?

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaReadSchemaTransformProvider.java [hunk 2]
```diff
@@ -41,19 +41,32 @@
...
     extends TypedSchemaTransformProvider<KafkaReadSchemaTransformConfiguration> {
 
+  final Boolean isTest;
+  final Integer testTimeoutSecs;
+
+  public KafkaReadSchemaTransformProvider() {
+    this(false, 0);
+  }
+
+  KafkaReadSchemaTransformProvider(Boolean isTest, Integer testTimeoutSecs) {
```
Comment (Abacn) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaReadSchemaTransformProvider.java:57]: Consider add a `@VisibleForTesting` annotation as this constructor is supposed to be used in test only?

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java [hunk 3]
```diff
@@ -502,6 +511,103 @@ public void testKafkaWithDelayedStopReadingFunction() {
...
+                            .addValue(new Random(numb).nextDouble())
+                            .addValue(
+                                Lists.newArrayList(
+                                    Long.toString(numb + 1),
+                                    Long.toString(numb + 2))) // childrenNames
+                            .build())
+                .collect(Collectors.toList()));
+
+    PipelineResult writeResult = writePipeline.run();
+    writeResult.waitUntilFinish();
```
Comment (Abacn) [inline sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java:605]: Shall we assert pipeline status (not fail)? Same for the read pipeline below.

**Review by Abacn — COMMENTED** [review]
Thanks! Left a couple of comments found.

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java (hunk 3 above)
Comment (pabloem) [reply sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java:635]: good idea. done!

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaReadSchemaTransformProvider.java (hunk 2 above)
Comment (pabloem) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaReadSchemaTransformProvider.java:59]: great point. done

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaReadSchemaTransformProvider.java (hunk 1 above)
Comment (pabloem) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaReadSchemaTransformProvider.java:86]: correct. Trying to standardize on lowercase

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java [hunk 4]
```diff
@@ -502,6 +511,103 @@ public void testKafkaWithDelayedStopReadingFunction() {
...
+          .addInt32Field("age")
+          .addBooleanField("ageIsEven")
+          .addDoubleField("temperature")
+          .addArrayField("childrenNames", Schema.FieldType.STRING)
+          .build();
+
+  private static final int FIVE_MINUTES_IN_MS = 5 * 60 * 1000;
+
+  @Test(timeout = FIVE_MINUTES_IN_MS)
+  public void testKafkaViaSchemaTransformJson() {}
```
[the remark this answers is not in the document]
Comment (pabloem) [reply sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java]: ahhh good catch. we don't have automatic conversion, but I've hardcoded the json schema

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+  @Override
+  public @UnknownKeyFor @NonNull @Initialized List<@UnknownKeyFor @NonNull @Initialized String>
+      outputCollectionNames() {
+    return Collections.emptyList();
+  }
+
+  @AutoValue
+  @DefaultSchema(AutoValueSchema.class)
+  public abstract static class KafkaWriteSchemaTransformConfiguration implements Serializable {
```
[the remark this answers is not in the document]
Comment (pabloem) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java:138]: changed it in the design doc.
we've decided to do [datastore][operation]schematransform

**Comment by pabloem** [comment]
Run Java PreCommit

**Comment by pabloem** [comment]
oops I'll look at kafka issues

**Comment by pabloem** [comment]
Run Java_Kafka_IO_Direct PreCommit

**Comment by Abacn** [comment]
Looks like need to add the topic `kafka-schema-transform` here: https://github.com/apache/beam/blob/6c24637b451d762ba39f54a56a8a335283173307/.test-infra/kubernetes/kafka-cluster/05-kafka/configmap-config.yaml#L35

**Comment by Abacn** [comment]
> Looks like need to add the topic `kafka-schema-transform` here:
> 
> https://github.com/apache/beam/blob/6c24637b451d762ba39f54a56a8a335283173307/.test-infra/kubernetes/kafka-cluster/05-kafka/configmap-config.yaml#L35

never mind, that line is for performance test.

**Comment by pabloem** [comment]
tests passed : )

**Review by Abacn — APPROVED** [review]
LGTM

Just curious how the last commit (https://github.com/apache/beam/pull/24495/commits/115fdbedda9bee05b8be102e66216b64c3ad86d7) fixed `Caused by: com.fasterxml.jackson.core.JsonParseException: Illegal character ((CTRL-CHAR, code 4)): only regular white space (\r, \n, \t) is allowed between tokens
 at [Source: (String)"\u000462||\u0001�7�\u000E\u0002Q�?\u0004\u000463\u000464\u0000"; line: 1, column: 2]`? In [here](https://ci-beam.apache.org/job/beam_PreCommit_Java_Kafka_IO_Direct_Phrase/21/testReport/junit/org.apache.beam.sdk.io.kafka/KafkaIOIT/testKafkaViaSchemaTransformJson_9/). Not obvious to me

**Comment by pabloem** [comment]
the same kafka instance was being used for all tests - what happened is that the JSON tests were running against the same topic, and thus they were re-reading data that previous tests had written. For some reason the data was corrupted.

**Pull request merged**